### PR TITLE
use kotlin lib

### PR DESCRIPTION
**Pull request by alisa911**

### CODE REVIEW

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2708

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/RoutingService.java [hunk 1]
```diff
@@ -349,36 +351,46 @@ private GPXUtilities.TrkSegment generateRouteSegments(List<RouteSegmentResult> r
...
 
         return trkSegment;
     }
 
+    public StringBundle toKotlinStringBundle(RouteDataBundle routeDataBundle) {
+        StringBundle kotlinBundle = new StringBundle();
+        for (Map.Entry<String, net.osmand.binary.StringBundle.Item<?>> entry : routeDataBundle.getMap().entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue().getValue() != null ? entry.getValue().getValue().toString() : null;
+            kotlinBundle.putString(key, value);
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/RoutingService.java:389]: Possibly, avoid putString() if value == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/GpxController.java [hunk 2]
```diff
@@ -97,22 +98,19 @@ public String clear(HttpServletRequest request, HttpSession httpSession) throws
...
 		return gson.toJson(Map.of("all", ctx.files));
 	}
 	
-	
-	
 	@GetMapping(path = {"/get-gpx-file"}, produces = "application/json")
 	@ResponseBody
-	public ResponseEntity<Resource> getGpx(@RequestParam(required = true) String name,
-			HttpSession httpSession) throws IOException {
+	public ResponseEntity<Resource> getGpx(@RequestParam String name, HttpSession httpSession) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/GpxController.java:108]: required = true or NPE (at least incorrect equals) at line 113 ?

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/GpxController.java [hunk 3]
```diff
@@ -128,28 +126,34 @@ public ResponseEntity<Resource> getGpx(@RequestParam(required = true) String nam
 	@PostMapping(path = {"/process-srtm"}, produces = "application/json")
 	public ResponseEntity<StreamingResponseBody> attachSrtm(@RequestPart(name = "file") @Valid @NotNull @NotEmpty MultipartFile file) throws IOException {
 		final StringBuilder err = new StringBuilder();
-		GPXFile gpxFile = GPXUtilities.loadGPXFile(file.getInputStream());
-		if (gpxFile.error != null) {
+		GpxFile gpxFile;
+		try (Source source = new Buffer().readFrom(file.getInputStream())) {
+			gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+		}
+		if (gpxFile.getError() != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/GpxController.java:133]: NPE if try {} failed

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/OsmGpxController.java [hunk 4]
```diff
@@ -159,7 +210,7 @@ private RouteFile getGpxFile(byte[] bytes, String idKey) {
 
 	private void addGeoDataToFeature(RouteFile file, Feature feature) {
 		GpxFile gpxFile = file.gpxFile;
-		List<WptPt> points = gpxFile.getAllPoints();
+		List<WptPt> points = gpxFile.getAllSegmentsPoints();
 		GpxTrackAnalysis analysis = file.analysis;
 		if (!points.isEmpty() && points.size() > 100 && analysis.getMaxDistanceBetweenPoints() < 1000) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/OsmGpxController.java:215]: Please reuse constants from DownloadOsmGPX

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java [hunk 5]
```diff
@@ -50,7 +51,7 @@ public ResponseEntity<String> deleteFav(@RequestBody String data,
                                             @RequestParam String fileName,
                                             @RequestParam Long updatetime) throws IOException {
         PremiumUserDevicesRepository.PremiumUserDevice dev = favoriteService.getUserId();
-        GPXFile file = favoriteService.createGpxFile(fileName, dev, updatetime);
+        GpxFile file = favoriteService.createGpxFile(fileName, dev, updatetime);
         if (file != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java:55]: Add && xxx.getError() == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java [hunk 6]
```diff
@@ -67,12 +68,12 @@ public ResponseEntity<String> updateAllFavorites(@RequestBody List<String> data,
                                                      @RequestParam Long updatetime,
                                                      @RequestParam boolean updateTimestamp) throws IOException {
         PremiumUserDevicesRepository.PremiumUserDevice dev = favoriteService.getUserId();
-        GPXFile file = favoriteService.createGpxFile(fileName, dev, updatetime);
+        GpxFile file = favoriteService.createGpxFile(fileName, dev, updatetime);
         UserdataService.ResponseFileStatus respNewGroup;
         if (file != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java:73]: Add && xxx.getError() == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java [hunk 7]
```diff
@@ -123,16 +124,16 @@ public ResponseEntity<String> updateFav(@RequestBody String data,
                                             @RequestParam Long newGroupUpdatetime,
                                             @RequestParam int ind) throws IOException {
         PremiumUserDevicesRepository.PremiumUserDevice dev = favoriteService.getUserId();
-        GPXUtilities.WptPt wptPt = webGpxParser.convertToWptPt(gson.fromJson(data, WebGpxParser.Wpt.class));
-        GPXFile newGpxFile = favoriteService.createGpxFile(newGroupName, dev, newGroupUpdatetime);
+        WptPt wptPt = webGpxParser.convertToWptPt(gson.fromJson(data, WebGpxParser.Wpt.class));
+        GpxFile newGpxFile = favoriteService.createGpxFile(newGroupName, dev, newGroupUpdatetime);
         if (newGpxFile != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java:129]: Add && xxx.getError() == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java [hunk 8]
```diff
@@ -123,16 +124,16 @@ public ResponseEntity<String> updateFav(@RequestBody String data,
...
         } else
             throw new OsmAndPublicApiException(UserdataService.ERROR_CODE_FILE_NOT_AVAILABLE,
                     UserdataService.ERROR_MESSAGE_FILE_IS_NOT_AVAILABLE);
         
         boolean changeGroup = !oldGroupName.equals(newGroupName);
-        GPXFile oldGpxFile = null;
+        GpxFile oldGpxFile = null;
         if (changeGroup) {
             oldGpxFile = favoriteService.createGpxFile(oldGroupName, dev, oldGroupUpdatetime);
             if (oldGpxFile != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java:139]: always != null, so add && xxx.getError() == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java [hunk 9]
```diff
@@ -175,11 +176,11 @@ public ResponseEntity<String> renameGroup(@RequestParam String oldName,
                                               @RequestParam String fullNewName,
                                               @RequestParam Long oldUpdatetime) throws IOException {
         PremiumUserDevicesRepository.PremiumUserDevice dev = favoriteService.getUserId();
-        GPXFile gpxFile = favoriteService.createGpxFile(fullOldName, dev, oldUpdatetime);
+        GpxFile gpxFile = favoriteService.createGpxFile(fullOldName, dev, oldUpdatetime);
         if (gpxFile != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/FavoriteController.java:180]: Add && xxx.getError() == null

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java [hunk 10]
```diff
@@ -387,28 +371,32 @@ public ResponseEntity<String> listFiles(@RequestParam(required = false) String n
...
 					InputStream in = uf.data != null ? new ByteArrayInputStream(uf.data)
 							: userdataService.getInputStream(uf);
 					if (in != null) {
-						GPXFile gpxFile = GPXUtilities.loadGPXFile(new GZIPInputStream(in));
-						if (gpxFile.error != null) {
+						GpxFile gpxFile;
+						try (Source source = new Buffer().readFrom(in)) {
+							gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+						}
+						if (gpxFile.getError() != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java:383]: NPE if try {} failed

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java [hunk 11]
```diff
@@ -558,12 +545,14 @@ public ResponseEntity<String> getGpxInfo(HttpServletResponse response, HttpServl
...
 			}
 			bin = userdataService.getInputStream(dev, userFile);
-
-			GPXFile gpxFile = GPXUtilities.loadGPXFile(new GZIPInputStream(bin));
-			if (gpxFile.error != null) {
+			GpxFile gpxFile;
+			try (Source source = new Buffer().readFrom(bin)) {
+				gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+			}
+			if (gpxFile.getError() != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java:552]: NPE if try {} failed

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java [hunk 12]
```diff
@@ -610,12 +598,15 @@ public ResponseEntity<String> getSrtmGpx(HttpServletResponse response, HttpServl
 				return ResponseEntity.ok(gson.toJson(Collections.singletonMap(INFO_KEY, userFile.details.get(SRTM_ANALYSIS))));
 			}
 			bin = userdataService.getInputStream(dev, userFile);
-			GPXFile gpxFile = GPXUtilities.loadGPXFile(new GZIPInputStream(bin));
-			if (gpxFile.error != null) {
+			GpxFile gpxFile;
+			try (Source source = new Buffer().readFrom(bin)) {
+				gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+			}
+			if (gpxFile.getError() != null) {
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java:605]: NPE if try {} failed

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/GpxService.java [hunk 13]
```diff
@@ -35,33 +38,33 @@ public class GpxService {
...
+        GpxTrackAnalysis analysis = getAnalysis(gpxFileForAnalyse, false);
         gpxData.analysis = webGpxParser.getTrackAnalysis(analysis, null);
         gpxData.pointsGroups = webGpxParser.getPointsGroups(gpxFileForAnalyse);
         if (analysis != null) {
             boolean addSpeed = analysis.getAvgSpeed() != 0.0 && !analysis.hasSpeedInTrack();
-            if (!gpxData.tracks.isEmpty() && (!analysis.pointAttributes.isEmpty() || addSpeed)) {
+            if (!gpxData.tracks.isEmpty() && (!analysis.getPointAttributes().isEmpty() || addSpeed)) {
                 webGpxParser.addAdditionalInfo(gpxData.tracks, analysis, addSpeed);
             }
         }
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/GpxService.java:61]: Please wrap it into `if (gpxFileForAnalyse.getError() == null)`

**Review by RZR-UA — CHANGES_REQUESTED** [review]
Finally,

1) merge the freshest master into your branch

2) deploy your branch on creator and test all features with "yarn test:test" (run full UI test)

File: java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java [hunk 14]
```diff
@@ -379,6 +379,10 @@ public ResponseEntity<String> listFiles(@RequestParam(required = false) String n
 						GpxFile gpxFile;
 						try (Source source = new Buffer().readFrom(in)) {
 							gpxFile = GpxUtilities.INSTANCE.loadGpxFile(source);
+						} catch (IOException e) {
+							LOG.error("Error reading gpx file: " + uf.name + " (id=" + uf.id + ") (userid=" + uf.userid + ")");
```
Comment (RZR-UA) [inline java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/user/MapApiController.java:383]: LOG.error("web-list-files: loadGpxFile error: " + uf.name + " (id=" + uf.id + ") (userid=" + uf.userid + ")");

please use similar keywords in message to make it "greppable" easier in the logs (grep "web-list-files" ...)

**Review by RZR-UA — APPROVED** [review]

**Review by RZR-UA — APPROVED** [review]

**Pull request merged**